tools/cart.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Convert an Atari 400/800/... cartridge rom to a CART file for the
+ * atari800 emulator (atari800.github.io)
+ * 
+ * See https://github.com/atari800/atari800/blob/master/DOC/cart.txt 
+ *
+ * 4 Jan 2020 - Walt Drummond <walt@drummond.us>
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <errno.h>
+#include <stdlib.h>
+
+#define MIN_CARTTYPE 1
+#define MAX_CARTTYPE 70
+#define URL "https://github.com/atari800/atari800/blob/master/DOC/cart.txt"
+
+uint32_t CARTRIDGE_Checksum(const uint8_t *image, int nbytes);
+long flen(FILE *fp);
+void convert(char *romfile, char *cartfile, int32_t type);
+
+typedef struct {
+  uint8_t cart[4];
+  uint8_t type[4];
+  uint8_t csum[4];
+  uint8_t zero[4];
+} header;
+
+/* Taken directly from atari800-4.1.0/src/cartridge.c */
+uint32_t CARTRIDGE_Checksum(const uint8_t *image, int nbytes)
+{
+	int checksum = 0;
+	while (nbytes > 0) {
+		checksum += *image++;
+		nbytes--;
+	}
+	return checksum;
+}
+
+long flen(FILE *fp)
+{
+  long l;
+  
+  fseek(fp, 0, SEEK_END);
+  l = ftell(fp);
+  rewind(fp);
+  return l;
+}
+
+void convert(char *romfile, char *cartfile, int32_t type)
+{
+  FILE *fp;
+  long len;
+  uint8_t *rom;
+  uint32_t csum;
+  header h;
+
+  printf("Converting %s to %s (type %d)\n", romfile, cartfile, type);
+
+  /* Read the original rom file */
+  fp = fopen(romfile, "rb");
+  if (fp == NULL) {
+    printf("Error opening rom file '%s': %s\n", romfile, strerror(errno));
+    exit(1);
+  }
+  len = flen(fp);
```

Comment:
You should check if the length is < 2048 or > 0x8000000 and show an error. Or better yet if you check that the size is compatible with the cartridge type given.

Reply:
Done.  Please take another look?